// maath/buffer/dist/maath-buffer.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('../../dist/objectSpread2-2ccd0bad.cjs.prod.js');
require('../../dist/triangle-9e5a8229.cjs.prod.js');
require('three');
require('../../dist/misc-8dab750e.cjs.prod.js');
require('../../dist/vector2-49e42f03.cjs.prod.js');
require('../../dist/vector3-4bbdb053.cjs.prod.js');
var buffer_dist_maathBuffer = require('../../dist/buffer-fd4cab21.cjs.prod.js');
require('../../dist/isNativeReflectConstruct-9acebf01.cjs.prod.js');
require('../../dist/matrix-e0b2acc5.cjs.prod.js');



exports.addAxis = buffer_dist_maathBuffer.addAxis;
exports.center = buffer_dist_maathBuffer.center;
exports.expand = buffer_dist_maathBuffer.expand;
exports.lerp = buffer_dist_maathBuffer.lerp;
exports.map = buffer_dist_maathBuffer.map;
exports.reduce = buffer_dist_maathBuffer.reduce;
exports.rotate = buffer_dist_maathBuffer.rotate;
exports.sort = buffer_dist_maathBuffer.sort;
exports.swizzle = buffer_dist_maathBuffer.swizzle;
exports.translate = buffer_dist_maathBuffer.translate;
